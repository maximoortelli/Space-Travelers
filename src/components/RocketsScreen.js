import { React, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRocketsAPI, reserveRocket, cancelReservation } from '../redux/rockets/RocketsSlice';
import RocketItem from './RocketItem';
import styles from '../styles/rocketsScreen.module.css';

const RocketsScreen = () => {
  const dispatch = useDispatch();
  const { rocketsData, isLoadingRockets, error } = useSelector((store) => store.rockets);
  const [rocketsReserved, setRocketsReserved] = useState([]);

  const handleClick = (rocketId) => {
    const isReserved = rocketsReserved.includes(rocketId);
    if (!isReserved) {
      dispatch(reserveRocket(rocketId));
      setRocketsReserved([...rocketsReserved, rocketId]);
    } else {
      dispatch(cancelReservation(rocketId));
      setRocketsReserved(rocketsReserved.filter((id) => rocketId !== id));
    }
  };

  useEffect(() => {
    if (isLoadingRockets && !rocketsData.length) {
      dispatch(fetchRocketsAPI());
    }
  }, [dispatch, isLoadingRockets, rocketsData.length]);

  if (isLoadingRockets) {
    return <div>Rockets loading...</div>;
  }

  if (error) {
    return <div>{`We encountered an error: ${JSON.stringify(error)}`}</div>;
  }

  if (!rocketsData.length) {
    return (
      <section>
        <header>
          <h2>There are no rockets for the moment</h2>
          <h4>Our rocket list is empty</h4>
        </header>
      </section>
    );
  }

  return (
    <div>
      <ul>
        {rocketsData.map((rocket) => (
          <li key={rocket.rocket_id} className={styles.rocketItemWrapper}>
            <RocketItem
              rocketId={rocket.rocket_id}
              rocketName={rocket.rocket_name}
              rocketDescription={rocket.description}
              flickrImages={rocket.flickr_images[0]}
              isReserved={rocket.isReserved} // Make sure it's passed
              onClick={handleClick}
            />
          </li>
        ))}
      </ul>
    </div>
  );
};

export default RocketsScreen;
