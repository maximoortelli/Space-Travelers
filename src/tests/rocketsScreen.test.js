import renderer from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux';
import { waitFor } from '@testing-library/react';
import axios from 'axios';
import configureStore from 'redux-mock-store';
import RocketsScreen from '../components/RocketsScreen';

jest.mock('axios');

describe('Confirm that RocketsScreen renders correctly when: ', () => {
  test('store is not empty', async () => {
    const mockData = [
      {
        rocket_id: '1',
        rocket_name: 'Falcon 9',
        description: 'A reusable two-stage rocket',
        flickr_images: ['https://example.com/falcon9.png'],
      },
      {
        rocket_id: '2',
        rocket_name: 'Starship',
        description: 'A fully reusable launch vehicle',
        flickr_images: ['https://example.com/starship.png'],
      },
    ];
    axios.get.mockResolvedValue({ data: mockData });

    const mockStore = configureStore();
    const initialState = {
      // Define your initial state here based on your reducers
      rockets: {
        rocketsData: mockData,
        isLoadingRockets: false,
        error: null,
      },
    };

    const store = mockStore(initialState);

    const component = (
      <Provider store={store}>
        <RocketsScreen />
      </Provider>
    );

    const container = renderer.create(component);

    const rocketScreenTree = container.toJSON();

    await waitFor(() => {
      expect(rocketScreenTree).toMatchSnapshot();
    });
  });

  test('store is empty', async () => {
    const mockData = [];
    axios.get.mockResolvedValue({ data: mockData });

    const mockStore = configureStore();
    const initialState = {
      // Define your initial state here based on your reducers
      rockets: {
        rocketsData: mockData,
        isLoadingRockets: false,
        error: null,
      },
    };

    const store = mockStore(initialState);

    const component = (
      <Provider store={store}>
        <RocketsScreen />
      </Provider>
    );

    const container = renderer.create(component);

    const rocketScreenTree = container.toJSON();

    await waitFor(() => {
      expect(rocketScreenTree).toMatchSnapshot();
    });
  });
  test('page is loading', async () => {
    const mockData = [];
    axios.get.mockResolvedValue({ data: mockData });

    const mockStore = configureStore();
    const initialState = {
      // Define your initial state here based on your reducers
      rockets: {
        rocketsData: mockData,
        isLoadingRockets: true,
        error: null,
      },
    };

    const store = mockStore(initialState);

    const component = (
      <Provider store={store}>
        <RocketsScreen />
      </Provider>
    );

    const container = renderer.create(component);

    const rocketScreenTree = container.toJSON();

    await waitFor(() => {
      expect(rocketScreenTree).toMatchSnapshot();
    });
  });
});
