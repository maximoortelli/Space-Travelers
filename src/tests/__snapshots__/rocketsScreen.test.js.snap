// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirm that RocketsScreen renders correctly when:  page is loading 1`] = `
<div>
  Rockets loading...
</div>
`;

exports[`Confirm that RocketsScreen renders correctly when:  store is empty 1`] = `
<section>
  <header>
    <h2>
      There are no rockets for the moment
    </h2>
    <h4>
      Our rocket list is empty
    </h4>
  </header>
</section>
`;

exports[`Confirm that RocketsScreen renders correctly when:  store is not empty 1`] = `
<div>
  <ul>
    <li
      className="rocketItemWrapper"
    >
      <div
        className="rocketItem"
      >
        <img
          alt="Falcon 9 rocket"
          className="rocketImg"
          src="https://example.com/falcon9.png"
        />
        <div
          className="rocketContent"
        >
          <h3
            className="rocketName"
          >
            Falcon 9
          </h3>
          <p
            className="rocketDescription"
          >
            A reusable two-stage rocket
          </p>
          <button
            className="reserveBtn"
            onClick={[Function]}
            type="button"
          >
            Reserve Rocket
          </button>
        </div>
      </div>
    </li>
    <li
      className="rocketItemWrapper"
    >
      <div
        className="rocketItem"
      >
        <img
          alt="Starship rocket"
          className="rocketImg"
          src="https://example.com/starship.png"
        />
        <div
          className="rocketContent"
        >
          <h3
            className="rocketName"
          >
            Starship
          </h3>
          <p
            className="rocketDescription"
          >
            A fully reusable launch vehicle
          </p>
          <button
            className="reserveBtn"
            onClick={[Function]}
            type="button"
          >
            Reserve Rocket
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
